File: KFC_Cpp/src/Board.cpp
#include "Board.hpp"

#include <cmath>    // std::round

// ---------------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------------
Board::Board(int cell_H_pix_,
             int cell_W_pix_,
             int W_cells_,
             int H_cells_,
             const Img& image,
             float cell_H_m_,
             float cell_W_m_)
    : cell_H_pix(cell_H_pix_),
      cell_W_pix(cell_W_pix_),
      W_cells(W_cells_),
      H_cells(H_cells_),
      img(image),
      cell_H_m(cell_H_m_),
      cell_W_m(cell_W_m_) {}

// ---------------------------------------------------------------------------
Board Board::clone() const {
    Img new_img = img;  // cv::Mat uses reference counting – shallow copy OK
    // If deep-copy is required uncomment: new_img = Img(img); but fine for now.
    return Board(cell_H_pix, cell_W_pix, W_cells, H_cells, new_img, cell_H_m, cell_W_m);
}

// ---------------------------------------------------------------------------
void Board::show() const {
    if (img.is_loaded()) {
        const_cast<Img&>(img).show();
    }
    // else: intentionally no-op – mirrors Python headless behaviour
}

// ---------------------------------------------------------------------------
std::pair<int, int> Board::m_to_cell(const std::pair<float, float>& pos_m) const {
    float x_m = pos_m.first;
    float y_m = pos_m.second;
    int col = static_cast<int>(std::round(x_m / cell_W_m));
    int row = static_cast<int>(std::round(y_m / cell_H_m));
    return {row, col};
}

std::pair<float, float> Board::cell_to_m(const std::pair<int, int>& cell) const {
    int r = cell.first;
    int c = cell.second;
    return {static_cast<float>(c) * cell_W_m, static_cast<float>(r) * cell_H_m};
}

std::pair<int, int> Board::m_to_pix(const std::pair<float, float>& pos_m) const {
    float x_m = pos_m.first;
    float y_m = pos_m.second;
    int x_px = static_cast<int>(std::round(x_m / cell_W_m * cell_W_pix));
    int y_px = static_cast<int>(std::round(y_m / cell_H_m * cell_H_pix));
    return {x_px, y_px};
} 

File: KFC_java/src/main/java/Board.java
public class Board {
    private final int cellHPix;
    private final int cellWPix;
    private final int wCells;
    private final int hCells;
    private final Img img;

    private final double cellHM;
    private final double cellWM;

    public Board(int cellHPix, int cellWPix,
                 int wCells, int hCells,
                 Img img) {
        this(cellHPix, cellWPix, wCells, hCells, img, 1.0, 1.0);
    }

    public Board(int cellHPix, int cellWPix,
                 int wCells, int hCells,
                 Img img,
                 double cellHM, double cellWM) {
        this.cellHPix = cellHPix;
        this.cellWPix = cellWPix;
        this.wCells = wCells;
        this.hCells = hCells;
        this.img = img;
        this.cellHM = cellHM;
        this.cellWM = cellWM;
    }

    /* ------------ convenience ------------- */

    public Board cloneBoard() {
        Img newImg;
        if (img != null && img.get() != null) {
            java.awt.image.BufferedImage copy = new java.awt.image.BufferedImage(
                    img.get().getColorModel(),
                    img.get().copyData(null),
                    img.get().isAlphaPremultiplied(),
                    null);
            BuffImg bi = new BuffImg();
            bi.setBufferedImage(copy);
            newImg = bi;
        } else {
            newImg = new BuffImg();
        }
        return new Board(cellHPix, cellWPix, wCells, hCells, newImg, cellHM, cellWM);
    }

    public void show() { if (img != null) img.show(); }

    /* ------------ conversions ------------- */

    /** Convert (x, y) in metres to board cell (row, col). */
    public int[] mToCell(double xM, double yM) {
        int col = (int) Math.round(xM / cellWM);
        int row = (int) Math.round(yM / cellHM);
        return new int[]{row, col};
    }

    public Moves.Pair mToCellPair(double xM, double yM) {
        int col = (int) Math.round(xM / cellWM);
        int row = (int) Math.round(yM / cellHM);
        return new Moves.Pair(row, col);
    }

    /** Convert a cell (row, col) to its top-left corner in metres. */
    public double[] cellToM(int row, int col) {
        return new double[]{col * cellWM, row * cellHM};
    }

    public double[] cellToM(Moves.Pair cell) {
        return cellToM(cell.r, cell.c);
    }

    /** Convert (x, y) in metres to pixel coordinates. */
    public int[] mToPix(double xM, double yM) {
        int xPx = (int) Math.round(xM / cellWM * cellWPix);
        int yPx = (int) Math.round(yM / cellHM * cellHPix);
        return new int[]{xPx, yPx};
    }

    /* ------------ getters ------------- */
    public int getCellHPix() { return cellHPix; }
    public int getCellWPix() { return cellWPix; }
    public int getWCells() { return wCells; }
    public int getHCells() { return hCells; }
    public Img getImg() { return img; }
} 